### PR TITLE
Support atomic updates and nested documents for JSON and XML serializations

### LINKED ISSUE #109: Why convert to XML? Support for nested docs?
If the documents are ruby hashes why convert it to XML, wouldn't be easier to convert to JSON?

Also, with this conversion is impossible to add nested documents. Using JSON, nested documents are added to "_childDocuments" field and Solr processes them automatically. With XML it's necessary to create a new doc and embed it into the parent.

### LINKED ISSUE #137: support for atomic updates?
I tried everything i could to build a RSrol:Xml with an atomic update with no success. if i send one nested hash, it converts the nested hash to string. 

Im running this:

```Ruby
RSolr::Xml::Document.new({ foo: { set: "Bar" } })
```

and the return:

```Ruby
#<RSolr::Xml::Document:0x007f940b61abb0 @fields=[#<RSolr::Xml::Field:0x007f940b61a840 @attrs={:name=>:foo}, @value="{:set=>\"Bar\"}">], @attrs={}>
```

Is there any way to build the correct xml with rsolr?

### PR DESCRIPTION
**Pull request by cbeer**
Fixes #137 
Fixes #109

### CODE REVIEW

File: lib/rsolr/document.rb [hunk 1]
```diff
@@ -38,7 +38,11 @@ def field_by_name(name)
     def add_field(name, values, options = {})
       wrap(values).each do |v|
         next if v.nil?
-        @fields << RSolr::Field.instance(options.merge({:name=>name}), v)
+        if name.to_s == '_childDocuments_'
```
Comment (jcoyne) [inline lib/rsolr/document.rb]: Should this stiring be frozen or should we add the frozen string pragma to the file?

File: lib/rsolr/document.rb [hunk 2]
```diff
@@ -68,9 +74,16 @@ def wrap(object)
...
 
     def self.instance(attrs, value)
-      field_type = attrs.fetch(:type, value.class.name) + "Field"
-      search_scope = Module.nesting[1]
-      klass = search_scope.const_defined?(field_type, false) ? search_scope.const_get(field_type) : Field
+      attrs = attrs.dup
+      field_type = attrs.delete(:type) {  value.class.name }
+
+      klass = if field_type.is_a? String
+        potential_class_name = field_type + 'Field'
```
Comment (jcoyne) [inline lib/rsolr/document.rb]: extract this logic to a new method.

File: lib/rsolr/document.rb [hunk 3]
```diff
@@ -124,4 +137,18 @@ def value
       source_value.to_time.getutc.iso8601
     end
   end
+
+  class DocumentField < Field
+    def value
+      if source_value.respond_to? :each_pair
```
Comment (jcoyne) [inline lib/rsolr/document.rb]: can we use the guard style here?

File: lib/rsolr/document.rb [hunk 4]
```diff
@@ -38,7 +38,11 @@ def field_by_name(name)
     def add_field(name, values, options = {})
       wrap(values).each do |v|
         next if v.nil?
-        @fields << RSolr::Field.instance(options.merge({:name=>name}), v)
+        if name.to_s == '_childDocuments_'
+          @fields << RSolr::Field.instance(options.merge({name: name, type: DocumentField }), v)
+        else
+          @fields << RSolr::Field.instance(options.merge({:name=>name}), v)
```
Comment (jcoyne) [inline lib/rsolr/document.rb]: These two lines are the same except `type:` in the hash. Can we put the hash construction in the conditional, and push the common part outside?

File: lib/rsolr/xml.rb [hunk 5]
```diff
@@ -75,16 +75,29 @@ def add data, add_attrs = nil, &block
...
-              end
-            end
+            doc_node_builder = to_xml(doc)
             self.class.use_nokogiri ? add_node.doc_(doc.attrs,&doc_node_builder) : add_node.doc(doc.attrs,&doc_node_builder)
           end
         end
       end
     end
+
+    def to_xml(doc)
```
Comment (jcoyne) [inline lib/rsolr/xml.rb]: Should this be private?

File: spec/api/document_spec.rb [hunk 6]
```diff
@@ -9,18 +9,20 @@
         expect(subject.instance({}, Time.new)).to be_a_kind_of(RSolr::TimeField)
       end
 
+      it "detect class reference by option" do
```
Comment (jcoyne) [inline spec/api/document_spec.rb]: "detects a"

File: spec/api/xml_spec.rb [hunk 7]
```diff
@@ -79,6 +79,17 @@
           expect(result).to match(/<field name="id">1<\/field>/)
         end
 
+        # add a single hash ("doc")
+        it 'should create an add from a hash formatted for atomic updates' do
```
Comment (jcoyne) [inline spec/api/xml_spec.rb]: Avoid using "should"

File: spec/api/xml_spec.rb [hunk 8]
```diff
@@ -79,6 +79,17 @@
...
+          data = {
+            :id=>1,
```
Comment (jcoyne) [inline spec/api/xml_spec.rb]: Use ruby 1.9 hash style (you already are on line 161)

File: lib/rsolr/document.rb (hunk 1 above)
Comment (cbeer) [inline lib/rsolr/document.rb]: 👌

File: spec/api/xml_spec.rb (hunk 7 above)
Comment (cbeer) [inline spec/api/xml_spec.rb]: I'm just blending in with the existing style; still want it changed?

File: lib/rsolr/document.rb (hunk 4 above)
Comment (cbeer) [inline lib/rsolr/document.rb]: 👌

File: lib/rsolr/document.rb (hunk 2 above)
Comment (cbeer) [inline lib/rsolr/document.rb]: 👌

File: lib/rsolr/xml.rb (hunk 5 above)
Comment (cbeer) [inline lib/rsolr/xml.rb]: 👌

File: lib/rsolr/document.rb (hunk 3 above)
Comment (cbeer) [inline lib/rsolr/document.rb]: 👌 . I'm not sure that's any more clear, so take a look.

File: spec/api/document_spec.rb (hunk 6 above)
Comment (cbeer) [inline spec/api/document_spec.rb]: 👌 , I've rewritten all those examples.

File: spec/api/xml_spec.rb (hunk 8 above)
Comment (jcoyne) [inline spec/api/xml_spec.rb]: Yep, I mainly want a consistent style. I don't care what style, just pick one and go with it.

Comment (cbeer) [inline spec/api/xml_spec.rb]: 👌 ; I've changed the new spec to use ruby 1.8-style hashes (as much as it pained me to do...)

**Pull request merged**